plugins/IPv4/plugin.go

```diff
@@ -0,0 +1,297 @@
+		if !ok {
+			log.Printf("plugins/IPv4: no new IP addresses available")
+			return resp, true
+
+		}
+		err := saveIPAddress(record, req.ClientHWAddr)
+		if err != nil {
+			log.Printf("plugins/IPv4: SaveIPAddress failed: %v", err)
+		}
+		StaticRecords[req.ClientHWAddr.String()] = record
+
+	}
+	ipaddr := record.IP
+	log.Printf("plugins/IPv4: found IP address %s for MAC %s", ipaddr, req.ClientHWAddr.String())
+	if req == nil {
+		log.Printf("plugins/IPv4: Packet is nil!")
+	}
+	if req.OpCode != dhcpv4.OpcodeBootRequest {
+		log.Printf("plugins/IPv4: Not a BootRequest!")
+	}
+	reply, err := dhcpv4.NewReplyFromRequest(req)
+	if err != nil {
+		log.Printf("plugins/IPv4: NewReplyFromRequest failed: %v", err)
+	}
+	switch mt := req.MessageType(); mt {
+	case dhcpv4.MessageTypeDiscover:
+		reply, err = dhcpv4.NewReplyFromRequest(req, dhcpv4.WithMessageType(dhcpv4.MessageTypeOffer), dhcpv4.WithYourIP(ipaddr), dhcpv4.WithNetmask(ClientSubnet), dhcpv4.WithRouter(serverIP), dhcpv4.WithDNS(DNSServer), dhcpv4.WithLeaseTime(LeaseTime))
+		if err != nil {
+			log.Printf("plugins/IPv4: NewReplyFromRequest failed: %v", err)
+		}
+		reply.UpdateOption(dhcpv4.OptServerIdentifier(serverIP))
+	case dhcpv4.MessageTypeRequest:
+		reply, err = dhcpv4.NewReplyFromRequest(req, dhcpv4.WithMessageType(dhcpv4.MessageTypeAck), dhcpv4.WithYourIP(ipaddr), dhcpv4.WithNetmask(ClientSubnet), dhcpv4.WithRouter(serverIP), dhcpv4.WithDNS(DNSServer), dhcpv4.WithLeaseTime(LeaseTime))
+		if err != nil {
+			log.Printf("plugins/IPv4: NewReplyFromRequest failed: %v", err)
+		}
+		reply.UpdateOption(dhcpv4.OptServerIdentifier(serverIP))
+	default:
+		log.Printf("plugins/IPv4: Unhandled message type: %v", mt)
+	}
+	return reply, true
+}
+
+// setupIPV6 not implemented only IPv4
+func setupIPV6(args ...string) (handler.Handler6, error) {
+	return nil, nil
+}
+
+func setupIPv4(args ...string) (handler.Handler4, error) {
+	_, h4, err := setupIP(false, args...)
+	return h4, err
+}
+
+func setupIP(v6 bool, args ...string) (handler.Handler6, handler.Handler4, error) {
+	if v6 {
+
+	} else {
+		if len(args) < 4 {
+			return nil, nil, errors.New("plugins/IPv4: need a file name, server IP, netmask and a DNS server")
+		}
+		filename = args[0]
+		if filename == "" {
+			return nil, nil, errors.New("plugins/IPv4: got empty file name")
+		}
+		serverIP = net.ParseIP(args[1])
+		if serverIP.To16() == nil {
+			return Handler6, Handler4, errors.New("plugins/IPv4: expected an IPv4 address, got: " + args[1])
+		}
+		netmask = net.ParseIP(args[2])
+		if netmask.To16() == nil {
+			return Handler6, Handler4, errors.New("plugins/IPv4: expected an IPv4 address, got: " + args[2])
+		}
+		if netmask.IsUnspecified() {
+			return Handler6, Handler4, errors.New("plugins/IPv4: netmask can not be 0.0.0.0, got: " + args[2])
+		}
+		if !checkValidNetmask(netmask) {
+			return Handler6, Handler4, errors.New("plugins/IPv4: netmask is not valid, got: " + args[2])
+		}
+		DNSServer = net.ParseIP(args[3])
```

Comment:
You are putting all the logic in one file. The idea of CoreDHCP is to make it as modular as possible.

In this case the best approach is to include the DNS plugin which will add the DNS option to the replies and remove all the DNS logic from here.

Reply:
Oh, I noticed there is no DNS plugin. One should be created.